File: libs/java/trinity-vsa/src/main/java/com/trinity/vsa/VSA.java
package com.trinity.vsa;

import java.util.List;

/**
 * Vector Symbolic Architecture operations
 */
public final class VSA {
    private VSA() {}

    /**
     * Bind two vectors (element-wise multiplication)
     */
    public static TritVector bind(TritVector a, TritVector b) {
        if (a.getDim() != b.getDim()) {
            throw new IllegalArgumentException("Dimension mismatch");
        }
        byte[] result = new byte[a.getDim()];
        byte[] aData = a.getData();
        byte[] bData = b.getData();
        
        for (int i = 0; i < a.getDim(); i++) {
            result[i] = (byte) (aData[i] * bData[i]);
        }
        return new TritVector(result);
    }

    /**
     * Unbind (inverse of bind, same operation for balanced ternary)
     */
    public static TritVector unbind(TritVector a, TritVector b) {
        return bind(a, b);
    }

    /**
     * Bundle multiple vectors via majority voting
     */
    public static TritVector bundle(List<TritVector> vectors) {
        if (vectors.isEmpty()) {
            throw new IllegalArgumentException("Empty vector list");
        }
        int dim = vectors.get(0).getDim();
        byte[] result = new byte[dim];
        
        for (int i = 0; i < dim; i++) {
            int sum = 0;
            for (TritVector v : vectors) {
                sum += v.get(i);
            }
            if (sum > 0) result[i] = 1;
            else if (sum < 0) result[i] = -1;
            else result[i] = 0;
        }
        return new TritVector(result);
    }

    /**
     * Bundle array of vectors
     */
    public static TritVector bundle(TritVector... vectors) {
        return bundle(List.of(vectors));
    }

    /**
     * Circular permutation
     */
    public static TritVector permute(TritVector v, int shift) {
        int dim = v.getDim();
        byte[] result = new byte[dim];
        byte[] data = v.getData();
        
        for (int i = 0; i < dim; i++) {
            int newIdx = Math.floorMod(i + shift, dim);
            result[newIdx] = data[i];
        }
        return new TritVector(result);
    }

    /**
     * Dot product
     */
    public static long dot(TritVector a, TritVector b) {
        if (a.getDim() != b.getDim()) {
            throw new IllegalArgumentException("Dimension mismatch");
        }
        long sum = 0;
        byte[] aData = a.getData();
        byte[] bData = b.getData();
        
        for (int i = 0; i < a.getDim(); i++) {
            sum += aData[i] * bData[i];
        }
        return sum;
    }

    /**
     * Cosine similarity
     */
    public static double similarity(TritVector a, TritVector b) {
        double d = dot(a, b);
        double normA = Math.sqrt(dot(a, a));
        double normB = Math.sqrt(dot(b, b));
        if (normA == 0 || normB == 0) return 0;
        return d / (normA * normB);
    }

    /**
     * Hamming distance
     */
    public static int hammingDistance(TritVector a, TritVector b) {
        if (a.getDim() != b.getDim()) {
            throw new IllegalArgumentException("Dimension mismatch");
        }
        int count = 0;
        byte[] aData = a.getData();
        byte[] bData = b.getData();
        
        for (int i = 0; i < a.getDim(); i++) {
            if (aData[i] != bData[i]) count++;
        }
        return count;
    }
}

File: src/libs/clojure/trinity-vsa/src/trinity/vsa.clj
(ns trinity.vsa
  "Trinity VSA - Vector Symbolic Architecture with balanced ternary arithmetic")

(defn zeros [dim]
  (vec (repeat dim 0)))

(defn random-vec [dim seed]
  (let [rng (java.util.Random. seed)]
    (vec (repeatedly dim #(- (.nextInt rng 3) 1)))))

(defn bind [a b]
  {:pre [(= (count a) (count b))]}
  (mapv * a b))

(defn unbind [a b]
  (bind a b))

(defn bundle [vectors]
  {:pre [(seq vectors)]}
  (let [dim (count (first vectors))]
    (vec (for [i (range dim)]
           (let [s (reduce + (map #(nth % i) vectors))]
             (cond (pos? s) 1 (neg? s) -1 :else 0))))))

(defn permute [v shift]
  (let [dim (count v)]
    (vec (for [i (range dim)]
           (nth v (mod (- i shift) dim))))))

(defn dot [a b]
  {:pre [(= (count a) (count b))]}
  (reduce + (map * a b)))

(defn similarity [a b]
  (let [d (double (dot a b))
        norm-a (Math/sqrt (double (dot a a)))
        norm-b (Math/sqrt (double (dot b b)))]
    (if (or (zero? norm-a) (zero? norm-b))
      0.0
      (/ d (* norm-a norm-b)))))

(defn hamming-distance [a b]
  {:pre [(= (count a) (count b))]}
  (count (filter true? (map not= a b))))
